ast/src/parsed/mod.rs

```diff
@@ -633,6 +635,71 @@ pub enum BinaryOperator {
     Greater,
 }
 
+#[derive(Debug, PartialEq, Eq)]
+pub enum BinaryOperatorAssociativity {
+    Left,
+    Right,
+    RequireParentheses,
+}
+
+const UNARY_EXPRESSION_PRIORITY_OFFSET: ExpressionPrecedence = 0; // All unary operators have the same precedence.
+const BINARY_EXPRESSION_PRIORITY_OFFSET: ExpressionPrecedence = 10; // Binary operators will have a lower precedence than unary operators.
+
+impl UnaryOperator {
+    pub fn precedence(&self) -> ExpressionPrecedence {
+        use UnaryOperator::*;
+        UNARY_EXPRESSION_PRIORITY_OFFSET
+            + match self {
+                Minus | LogicalNot => 1,
+                Next => 2,
+            }
+    }
+}
+
+impl BinaryOperator {
+    pub fn precedence(&self) -> ExpressionPrecedence {
+        use BinaryOperator::*;
+        BINARY_EXPRESSION_PRIORITY_OFFSET
+            + match self {
+                // Unary - * ! & &mut
+                Pow => 2,
+                // * / %
+                Mul | Div | Mod => 3,
+                // + -
+                Add | Sub => 4,
+                // << >>
+                ShiftLeft | ShiftRight => 5,
+                // &
+                BinaryAnd => 6,
+                // ^
+                BinaryXor => 7,
+                // |
+                BinaryOr => 8,
+                // == != < > <= >=
+                Equal | NotEqual | Less | Greater | LessEqual | GreaterEqual => 9,
+                // &&
+                LogicalAnd => 10,
+                // ||
+                LogicalOr => 11,
+                // .. ..=
+                // ??
+                // = += -= *= /= %= &= |= ^= <<= >>=
+                Identity => 12,
+            }
+    }
+
+    pub fn associativity(&self) -> BinaryOperatorAssociativity {
+        use BinaryOperator::*;
+        use BinaryOperatorAssociativity::*;
+        match self {
+            Identity => Right,
+            Equal | NotEqual | Less | Greater | LessEqual | GreaterEqual => RequireParentheses,
+            // .. ..= => RequireParentheses,
+            _ => Left,
```

Comment:
What about Pow? Is it correct that it is left-associative?

Reply:
I found something in polygon pil that says it's left associative if I am not mistaken 
>>> I found this in pilcom source code, I assume power is left assosiative and we must do the same
https://github.com/0xPolygonHermez/pilcom/blob/36bfb6ab4e0f1bae3bbeda11cc6305ede153f6b6/src/pil_parser.jison#L51
@chriseth can you double check?

Can you double check?

Reply:
I think exponentiation only makes sense as right-associative. It is fine to break compatibility with polygon-pil there. How would you understand `x**2**16`? And please double-check either with the grammar or how it is parsed on an example.